File: simplified-books-core-ng/src/main/java/org/nypl/simplified/books/controller/ProfileAccountDeleteTask.java
package org.nypl.simplified.books.controller;

import com.io7m.jfunctional.FunctionType;
import com.io7m.jfunctional.Option;
import com.io7m.jfunctional.Unit;
import com.io7m.jnull.NullCheck;

import org.nypl.simplified.books.accounts.AccountEvent;
import org.nypl.simplified.books.accounts.AccountEventDeletion;
import org.nypl.simplified.books.accounts.AccountEventDeletion.AccountDeletionFailed;
import org.nypl.simplified.books.accounts.AccountEventDeletion.AccountDeletionSucceeded;
import org.nypl.simplified.books.accounts.AccountProvider;
import org.nypl.simplified.books.accounts.AccountProviderCollection;
import org.nypl.simplified.books.accounts.AccountType;
import org.nypl.simplified.books.accounts.AccountsDatabaseLastAccountException;
import org.nypl.simplified.books.profiles.ProfileAccountSelectEvent.ProfileAccountSelectSucceeded;
import org.nypl.simplified.books.profiles.ProfileEvent;
import org.nypl.simplified.books.profiles.ProfileNonexistentAccountProviderException;
import org.nypl.simplified.books.profiles.ProfileType;
import org.nypl.simplified.books.profiles.ProfilesDatabaseType;
import org.nypl.simplified.observable.ObservableType;

import java.net.URI;
import java.util.concurrent.Callable;

import static org.nypl.simplified.books.accounts.AccountEventDeletion.AccountDeletionFailed.ErrorCode.ERROR_ACCOUNT_ONLY_ONE_REMAINING;
import static org.nypl.simplified.books.accounts.AccountEventDeletion.AccountDeletionFailed.ErrorCode.ERROR_GENERAL;

final class ProfileAccountDeleteTask implements Callable<AccountEventDeletion> {

  private final ProfilesDatabaseType profiles;
  private final FunctionType<Unit, AccountProviderCollection> account_providers;
  private final URI provider_id;
  private final ObservableType<ProfileEvent> profile_events;
  private final ObservableType<AccountEvent> account_events;

  ProfileAccountDeleteTask(
    final ProfilesDatabaseType profiles,
    final ObservableType<AccountEvent> account_events,
    final ObservableType<ProfileEvent> profile_events,
    final FunctionType<Unit, AccountProviderCollection> account_providers,
    final URI provider) {

    this.profiles =
      NullCheck.notNull(profiles, "Profiles");
    this.account_events =
      NullCheck.notNull(account_events, "Account events");
    this.profile_events =
      NullCheck.notNull(profile_events, "Profile events");
    this.account_providers =
      NullCheck.notNull(account_providers, "Account providers");
    this.provider_id =
      NullCheck.notNull(provider, "Provider");
  }

  protected AccountEventDeletion execute() {

    try {
      final AccountProviderCollection providers_now = this.account_providers.call(Unit.unit());
      final AccountProvider provider = providers_now.providers().get(this.provider_id);

      if (provider != null) {
        final ProfileType profile = this.profiles.currentProfileUnsafe();
        final AccountType account_then = profile.accountCurrent();
        profile.deleteAccountByProvider(provider);
        final AccountType account_now = profile.accountCurrent();
        if (!account_now.id().equals(account_then.id())) {
          this.profile_events.send(
            ProfileAccountSelectSucceeded.of(account_then.id(), account_now.id()));
        }
        return AccountDeletionSucceeded.of(account_now.id(), provider);
      }

      throw new ProfileNonexistentAccountProviderException("Unrecognized provider: " + this.provider_id);
    } catch (final AccountsDatabaseLastAccountException e) {
      return AccountDeletionFailed.of(ERROR_ACCOUNT_ONLY_ONE_REMAINING, Option.some(e));
    } catch (final Exception e) {
      return AccountDeletionFailed.of(ERROR_GENERAL, Option.some(e));
    }
  }

  @Override
  public AccountEventDeletion call() throws Exception {
    final AccountEventDeletion event = execute();
    this.account_events.send(event);
    return event;
  }
}

File: simplified-books-controller/src/main/java/org/nypl/simplified/books/controller/ProfileAccountDeleteTask.kt
package org.nypl.simplified.books.controller

import io.reactivex.subjects.Subject
import org.nypl.simplified.accounts.api.AccountEvent
import org.nypl.simplified.accounts.api.AccountEventDeletion
import org.nypl.simplified.accounts.api.AccountID
import org.nypl.simplified.accounts.database.api.AccountsDatabaseLastAccountException
import org.nypl.simplified.metrics.api.MetricEvent
import org.nypl.simplified.metrics.api.MetricServiceType
import org.nypl.simplified.profiles.api.ProfileEvent
import org.nypl.simplified.profiles.api.ProfilesDatabaseType
import org.nypl.simplified.profiles.controller.api.ProfileAccountDeletionStringResourcesType
import org.nypl.simplified.taskrecorder.api.TaskRecorder
import org.nypl.simplified.taskrecorder.api.TaskResult
import org.nypl.simplified.taskrecorder.api.TaskStep
import org.slf4j.LoggerFactory
import java.net.URI
import java.util.concurrent.Callable

class ProfileAccountDeleteTask(
  private val accountEvents: Subject<AccountEvent>,
  private val accountProviderID: URI,
  private val profiles: ProfilesDatabaseType,
  private val profileEvents: Subject<ProfileEvent>,
  private val strings: ProfileAccountDeletionStringResourcesType,
  private val metrics: MetricServiceType?
) : Callable<TaskResult<Unit>> {

  private val logger = LoggerFactory.getLogger(ProfileAccountDeleteTask::class.java)
  private val taskRecorder = TaskRecorder.create()

  private fun publishFailureEvent() =
    this.accountEvents.onNext(
      AccountEventDeletion.AccountEventDeletionFailed(
        this.taskRecorder.finishFailure<Unit>()
      )
    )

  private fun publishProgressEvent(step: TaskStep) =
    this.accountEvents.onNext(AccountEventDeletion.AccountEventDeletionInProgress(step.description))

  private fun publishSuccessEvent(accountThen: AccountID) =
    this.accountEvents.onNext(
      AccountEventDeletion.AccountEventDeletionSucceeded(
        this.strings.deletionSucceeded, accountThen
      )
    )

  override fun call(): TaskResult<Unit> {
    return try {
      this.logger.debug("deleting account for provider {}", this.accountProviderID)
      this.publishProgressEvent(this.taskRecorder.beginNewStep(this.strings.deletingAccount))

      val profile = this.profiles.currentProfileUnsafe()
      val account = profile.deleteAccountByProvider(this.accountProviderID)

      this.metrics?.logMetric(MetricEvent.LibraryRemoved(this.accountProviderID.toString()))
      this.publishSuccessEvent(account)
      this.taskRecorder.finishSuccess(Unit)
    } catch (e: AccountsDatabaseLastAccountException) {
      this.publishFailureEvent()
      this.taskRecorder.finishFailure()
    } catch (e: Throwable) {
      this.publishFailureEvent()
      this.taskRecorder.finishFailure()
    }
  }
}
